File: bindings/go/dstu/box512_test.go
package dstu

import (
	"bytes"
	"testing"
)

// Box512SecretKey/Box512PublicKey (crypto_box512) - l(p)=512 sibling of BoxSecretKey/BoxPublicKey
// (T-193/T-204). No official vector exists for this composite construction (same posture as
// crypto_box) - correctness (round trip), rejection (tampered wire segments, wrong key), misuse
// (wrong-length/invalid key encodings, truncated input).

func TestBox512SealOpenRoundTrips(t *testing.T) {
	secretKey, _ := GenerateBox512SecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	message := []byte("a message for the public key's holder only")
	sealed, err := publicKey.Seal(message)
	if err != nil {
		t.Fatal(err)
	}
	got, err := secretKey.Open(sealed)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(got, message) {
		t.Fatalf("got %q, want %q", got, message)
	}
}

func TestBox512SealHandlesEmptyMessage(t *testing.T) {
	secretKey, _ := GenerateBox512SecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	sealed, err := publicKey.Seal(nil)
	if err != nil {
		t.Fatal(err)
	}
	got, err := secretKey.Open(sealed)
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 0 {
		t.Fatalf("got %d bytes, want 0", len(got))
	}
}

func TestBox512TwoSealsUseDifferentEphemeralMaterial(t *testing.T) {
	secretKey, _ := GenerateBox512SecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	message := []byte("same message twice")
	a, _ := publicKey.Seal(message)
	b, _ := publicKey.Seal(message)
	if bytes.Equal(a, b) {
		t.Fatal("two seals produced identical output")
	}
}

func TestBox512TamperedCiphertextIsRejected(t *testing.T) {
	secretKey, _ := GenerateBox512SecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	sealed, _ := publicKey.Seal([]byte("message"))
	sealed[len(sealed)-1] ^= 1
	if _, err := secretKey.Open(sealed); err == nil {
		t.Fatal("expected an error")
	}
}

func TestBox512WrongSecretKeyIsRejected(t *testing.T) {
	secretKey, _ := GenerateBox512SecretKey()
	defer secretKey.Close()
	publicKey := secretKey.PublicKey()
	defer publicKey.Close()
	otherSecretKey, _ := GenerateBox512SecretKey()
	defer otherSecretKey.Close()
	sealed, _ := publicKey.Seal([]byte("message"))
	if _, err := otherSecretKey.Open(sealed); err == nil {
		t.Fatal("expected an error")
	}
}

func TestBox512WrongLengthSecretKeyIsRejected(t *testing.T) {
	if _, err := Box512SecretKeyFromBytes(make([]byte, 10)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestBox512ZeroSecretKeyIsRejected(t *testing.T) {
	if _, err := Box512SecretKeyFromBytes(make([]byte, Box512SecretKeyBytes)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestBox512WrongLengthPublicKeyIsRejected(t *testing.T) {
	if _, err := Box512PublicKeyFromBytes(make([]byte, 10)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestBox512DegeneratePublicKeyXIsRejected(t *testing.T) {
	if _, err := Box512PublicKeyFromBytes(make([]byte, Box512PublicKeyBytes)); err == nil { // x = 0
		t.Fatal("expected an error")
	}
}

func TestBox512TruncatedSealedInputIsRejected(t *testing.T) {
	secretKey, _ := GenerateBox512SecretKey()
	defer secretKey.Close()
	if _, err := secretKey.Open([]byte("short")); err == nil {
		t.Fatal("expected an error")
	}
}

File: bindings/nodejs/test/box512.test.js
'use strict';

/**
 * `crypto_box512` - `l(p)=512` sibling of `crypto_box` (T-193/T-204). No official vector exists
 * for this composite construction (same posture as `crypto_box`) - correctness (round trip),
 * rejection (tampered wire segments, wrong key), misuse (wrong-length/invalid key encodings,
 * truncated input). Mirrors bindings/python/tests/test_box512.py.
 */

const test = require('node:test');
const assert = require('node:assert/strict');
const dstu = require('../js/index.js');

test('seal/open round-trips', () => {
  const secretKey = dstu.box512Keygen();
  const publicKey = dstu.box512PublicKey(secretKey);
  const message = Buffer.from("a message for the public key's holder only");
  const sealed = dstu.box512Seal(publicKey, message);
  assert.deepStrictEqual(dstu.box512Open(secretKey, sealed), message);
});

test('seal handles empty message', () => {
  const secretKey = dstu.box512Keygen();
  const publicKey = dstu.box512PublicKey(secretKey);
  const sealed = dstu.box512Seal(publicKey, Buffer.alloc(0));
  assert.deepStrictEqual(dstu.box512Open(secretKey, sealed), Buffer.alloc(0));
});

test('two seals use different ephemeral material', () => {
  const secretKey = dstu.box512Keygen();
  const publicKey = dstu.box512PublicKey(secretKey);
  const message = Buffer.from('same message twice');
  assert.notDeepStrictEqual(dstu.box512Seal(publicKey, message), dstu.box512Seal(publicKey, message));
});

test('tampered ciphertext is rejected', () => {
  const secretKey = dstu.box512Keygen();
  const publicKey = dstu.box512PublicKey(secretKey);
  const sealed = Buffer.from(dstu.box512Seal(publicKey, Buffer.from('message')));
  sealed[sealed.length - 1] ^= 1;
  assert.throws(() => dstu.box512Open(secretKey, sealed));
});

test('wrong secret key is rejected', () => {
  const secretKey = dstu.box512Keygen();
  const publicKey = dstu.box512PublicKey(secretKey);
  const otherSecretKey = dstu.box512Keygen();
  const sealed = dstu.box512Seal(publicKey, Buffer.from('message'));
  assert.throws(() => dstu.box512Open(otherSecretKey, sealed));
});

test('wrong-length secret key is rejected', () => {
  assert.throws(() => dstu.box512PublicKey(Buffer.from('too short')));
});

test('zero secret key is rejected', () => {
  assert.throws(() => dstu.box512PublicKey(Buffer.alloc(64)));
});

test('wrong-length public key is rejected', () => {
  assert.throws(() => dstu.box512Seal(Buffer.from('too short'), Buffer.from('message')));
});

test('degenerate public key x is rejected', () => {
  assert.throws(() => dstu.box512Seal(Buffer.alloc(64), Buffer.from('message'))); // x = 0
});

test('truncated sealed input is rejected', () => {
  const secretKey = dstu.box512Keygen();
  assert.throws(() => dstu.box512Open(secretKey, Buffer.from('short')));
});
